destroy: provide a way to stop all uninstalls using context

staebler opened the pull request:
The Run function in the Destroyer interface was modified to take a context as a parameter. This provides a way for the user stop the uninstall after a period of time by providing a context with a deadline.

The baremetal, libvirt, openstack, and ovirt providers do not provide a means by which most requests made to the provider can be stopped prematurely. In these cases, the context is checked prior to making the requests as a best effort. But the uninstall may continue for a period of time after the context is done.

The RunWithContext function introduced in https://github.com/openshift/installer/pull/3765 for AWS has been obsoleted since the Run function now accepts a context.

This will be used by Hive to backoff uninstall attempts.

https://issues.redhat.com/browse/CO-974

abhinavdahiya commented:
> A "--timeout" flag was added to the openshift-install destroy cluster command. This allows the user to time out the destroy after a specified number of seconds.

If the users want to add a timeout, they should invoke the installer with `timeout openshift-install ...`
I don't think the installer destroy should provide that api.

abhinavdahiya commented:
also if you can maybe split these into per-platform, switch destroy to pass context that would be very helpful in reviewing.

staebler commented:
Update to put the changes for each provider in separate commits and to remove the "--timeout" flag.

abhinavdahiya commented:
> https://github.com/openshift/installer/pull/3800/commits/019575162a4fb72c4c44ea6aeb3737033284675d

So i think GCP needs a relook. there is a global context for Uninstaller that is used to create new contexts with tighter timeouts for requests. This change just overwrites that as part of Run, while also removing the local tigher context for API calls.

> https://github.com/openshift/installer/pull/3800/commits/1685b1f92c45e47a72c579c19ea8c03144379972

hmm.. the placement to the ctx Errs seems a little random to me, is it possible to provide a little more context in the commit message? setting the convention like check for ctx before each item to be deleted or maybe only once per cycle etc. is what we need to try should help guide if change is good.

> https://github.com/openshift/installer/pull/3800/commits/ed8ccf3d475c086f61b3ef43250072d0c1010302

this needs review from openstack folks, cc @mandre @Fedosin

abhinavdahiya commented on pkg/destroy/ovirt/destroyer.go:
```diff
@@ -53,20 +56,34 @@ func (uninstaller *ClusterUninstaller) removeVMs(con *ovirtsdk.Connection) error
 	}
 	wg.Wait()
-	return nil
+	select {
+	case err := <-errChan:
+		return err
+	default:
+		return nil
+	}
```
why not collection all the errors instead of first one?

staebler replied:
I changed this to aggregate all of the errors.

staebler commented:
> > [0195751](https://github.com/openshift/installer/commit/019575162a4fb72c4c44ea6aeb3737033284675d)
> 
> So i think GCP needs a relook. there is a global context for Uninstaller that is used to create new contexts with tighter timeouts for requests. This change just overwrites that as part of Run, while also removing the local tigher context for API calls.

I fixed this to use the 2-minute timeout for the calls to get the session and services.

> > [1685b1f](https://github.com/openshift/installer/commit/1685b1f92c45e47a72c579c19ea8c03144379972)
> 
> hmm.. the placement to the ctx Errs seems a little random to me, is it possible to provide a little more context in the commit message? setting the convention like check for ctx before each item to be deleted or maybe only once per cycle etc. is what we need to try should help guide if change is good.

I fixed this to check the context directly before every request.

mandre commented:
/test e2e-openstack

staebler commented:
78e20f8a6...a33077627
* In gcp destroy, change `PollImmediateInfinite` to `PollImmediateUntil` to stop the polling as soon as the context is done rather than waiting for the polling interval to elapse.

mandre commented on pkg/destroy/openstack/openstack.go:
```diff
@@ -97,52 +98,63 @@ func (o *ClusterUninstaller) Run(context.Context) error {
+		wg.Add(1)
+		go func(name string, function deleteFunc) {
+			defer wg.Done()
+			err := deleteRunner(ctx, function, opts, o.Filter, o.Logger.WithField("deleteFuncName", name))
```
is the `deleteFuncName` here to stay? The output feels a bit "loaded".

```
...
DEBUG Deleting openstack security-groups            deleteFuncName=deleteSecurityGroups
DEBUG Deleting openstack networks                   deleteFuncName=deleteNetworks
DEBUG Deleting Port "f5f45370-10ed-421c-82f9-7c771482acc7"  deleteFuncName=deletePorts
DEBUG Deleting Subnet: "87507b1d-6be9-4fee-a560-84d87ff3ae93"  deleteFuncName=deleteSubnets
DEBUG Deleting network: "95264d20-2c03-4f95-84c6-f6faa8e8c6af"  deleteFuncName=deleteNetworks
DEBUG Deleting Security Group: "04fbdf3c-f3e5-43b8-b9b0-801fb4d167ae"  deleteFuncName=deleteSecurityGroups
DEBUG Exiting deleting openstack ports              deleteFuncName=deletePorts
DEBUG Deleting Security Group: "ab3903e2-e659-4253-9445-70cf351a26e4"  deleteFuncName=deleteSecurityGroups
DEBUG Deleting openstack trunks                     deleteFuncName=deleteTrunks
DEBUG Exiting deleting openstack security-groups    deleteFuncName=deleteSecurityGroups
DEBUG Exiting deleting openstack subnets            deleteFuncName=deleteSubnets
...
```

staebler replied:
I am fine with removing it if you do not want it. I added it to preserve the info that had been logged in the "goroutine %v complete" debug entry.

mandre commented on pkg/destroy/openstack/openstack.go:
```diff
@@ -97,52 +98,63 @@ func (o *ClusterUninstaller) Run(context.Context) error {
-	// record that the goroutine has run to completion
-	channel <- deleteFuncName
-	return
+	logger.Debugf("delete func complete")
```
We don't need this I think, every delete function logs when exiting.

staebler replied:
Removing it is fine with me. It was added to replace the "goroutine {func_name} complete" debug entry that I removed.

pierreprinetti commented:
/test e2e-openstack

pierreprinetti commented:
/retest

jstuever commented:
gcp changes looks good to me.

staebler commented:
a33077627...783ba77d2
* Rebased onto master
* Removed [log field](https://github.com/openshift/installer/pull/3800#discussion_r447650329) and [log entry](https://github.com/openshift/installer/pull/3800#discussion_r447652862) from OpenStack destroyer.

pierreprinetti commented:
/retest

pierreprinetti commented:
openstack-e2e is passing now.
